transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,415 @@
+    new_addrs: RwLock<HashSet<SocketAddr>>,
+}
+
+impl UDPMuxNewAddr {
+    pub fn new(params: UDPMuxParams, new_addr_tx: mpsc::Sender<Multiaddr>) -> Arc<Self> {
+        let (closed_watch_tx, closed_watch_rx) = watch::channel(());
+
+        let mux = Arc::new(Self {
+            params,
+            conns: Mutex::default(),
+            address_map: RwLock::default(),
+            closed_watch_tx: Mutex::new(Some(closed_watch_tx)),
+            closed_watch_rx: closed_watch_rx.clone(),
+            new_addrs: RwLock::default(),
+        });
+
+        let cloned_mux = Arc::clone(&mux);
+        cloned_mux.start_conn_worker(closed_watch_rx, new_addr_tx);
+
+        mux
+    }
+
+    pub async fn is_closed(&self) -> bool {
+        self.closed_watch_tx.lock().await.is_none()
+    }
+
+    async fn send_to(&self, buf: &[u8], target: &SocketAddr) -> Result<usize, Error> {
+        self.params
+            .conn
+            .send_to(buf, *target)
+            .await
+            .map_err(Into::into)
+    }
+
+    /// Create a muxed connection for a given ufrag.
+    async fn create_muxed_conn(self: &Arc<Self>, ufrag: &str) -> Result<UDPMuxConn, Error> {
+        let local_addr = self.params.conn.local_addr().await?;
+
+        let params = UDPMuxConnParams {
+            local_addr,
+            key: ufrag.into(),
+            udp_mux: Arc::clone(self),
+        };
+
+        Ok(UDPMuxConn::new(params))
+    }
+
+    async fn register_conn_for_address(&self, conn: &UDPMuxConn, addr: SocketAddr) {
+        if self.is_closed().await {
+            return;
+        }
+
+        let key = conn.key();
+        {
+            let mut addresses = self.address_map.write();
+
+            addresses
+                .entry(addr)
+                .and_modify(|e| {
+                    if e.key() != key {
+                        e.remove_address(&addr);
+                        *e = conn.clone()
+                    }
+                })
+                .or_insert_with(|| conn.clone());
+        }
+
+        // remove addr from new_addrs once conn is established
+        {
+            let mut new_addrs = self.new_addrs.write();
+            new_addrs.remove(&addr);
+        }
+
+        log::debug!("Registered {} for {}", addr, key);
+    }
+
+    async fn conn_from_stun_message(&self, buffer: &[u8], addr: &SocketAddr) -> Option<UDPMuxConn> {
+        match ufrag_from_stun_message(buffer, true) {
+            Ok(ufrag) => {
+                let conns = self.conns.lock().await;
+                conns.get(&ufrag).map(Clone::clone)
+            },
+            Err(e) => {
+                log::error!("{} (addr: {})", e, addr);
+                None
+            },
+        }
+    }
+
+    fn start_conn_worker(
+        self: Arc<Self>,
+        mut closed_watch_rx: watch::Receiver<()>,
+        mut new_addr_tx: mpsc::Sender<Multiaddr>,
+    ) {
+        tokio::spawn(async move {
+            let mut buffer = [0u8; RECEIVE_MTU];
+
+            loop {
```

Comment:
I am not too happy about spawning tasks with endless loops from within here. With https://github.com/libp2p/rust-libp2p/pull/2652, we will be getting a `poll` function on `Transport`. This would allow us to put this worker as a field into the actual `Transport` and continuously make progress by polling it.

Doing that may perhaps also allow us to remove all the `Arc`-ing from within here? That would be much closer to how the rest of `rust-libp2p` is designed.

Reply:
What is your rough timeline on this PR? Judging from just the patched dependencies for example, it seems that we are not going to merge this tomorrow, meaning we should be able to wait until #2652 lands and use the `Transport`s poll mechanism.